oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp

```diff
@@ -0,0 +1,280 @@
+                         (std::string_view::npos != msg.find(" / Password ")) &&
+                         (std::string_view::npos !=
+                          msg.find(" exceeds IPMI 16/20 limit")))
+                {
+                    // Handle messages from
+                    // https://github.com/openbmc/pam-ipmi/blob/master/src/pam_ipmicheck/pam_ipmicheck.c
+                    pfw_sms_appdata_ptr->reasonCode =
+                        PASSWORD_TOO_LONG_FOR_IPMI_USER;
+                }
+                // Ignore other messges, such as: You are required to change
+                // your password immediately (administrator enforced).
+                break;
+            default:
+                // Should never get here
+                return PAM_CONV_ERR;
+                break;
+        }
+    }
+    return PAM_SUCCESS;
+}
+
+/** @brief Common code factored out between authenticate() and changePassword().
+ */
+static void common_authentication_handler(const std::string& username,
+                                          const std::string& password,
+                                          bool& authenticated,
+                                          bool& passwordChangeRequired,
+                                          bool& operationAllowed,
+                                          bool& passwordChangeAllowed)
+{
+    authenticated = false; // default to fail safe (no access)
+    passwordChangeRequired = false;
+    operationAllowed = false;
+    passwordChangeAllowed = false;
+
+    struct pfw_sms_pam_appdata appdata = {
+        password.c_str(), PASSWORD_CHANGE_FAILED_UNKNOWN_REASON};
+    char* appData = reinterpret_cast<char*>(&appdata);
+    const struct pam_conv pamConversation = {pamConversationFunction, appData};
+    pam_handle_t* pamHandle = nullptr; // this gets set by pam_start
+    int retval = pam_start("ibm-pfw-sms-menu", username.c_str(),
+                           &pamConversation, &pamHandle);
+    if (retval != PAM_SUCCESS)
+    {
+        return;
+    }
+
+    retval = pam_authenticate(pamHandle,
+                              PAM_SILENT | PAM_DISALLOW_NULL_AUTHTOK);
+    if (retval != PAM_SUCCESS)
+    {
+        pam_end(pamHandle, retval); // ignore retval
+        return;
+    }
+
+    authenticated = true;
+
+    /* check that the account is healthy */
+    retval = pam_acct_mgmt(pamHandle, PAM_DISALLOW_NULL_AUTHTOK);
+    if (retval == PAM_NEW_AUTHTOK_REQD)
+    {
+        operationAllowed = false;
+        passwordChangeRequired = true;
+        passwordChangeAllowed = true;
+        pam_end(pamHandle, retval); // ignore retval
+        return;
+    }
+    if (retval != PAM_SUCCESS)
+    {
+        pam_end(pamHandle, retval); // ignore retval
+        return;                     // operationAllowed == false
+    }
+    operationAllowed = true;
+    passwordChangeAllowed = true;
+
+    retval = pam_end(pamHandle, retval); // ignore retval
```

Comment:
assigning of retval won't make any sense since it's last line and retval is local scope & function returns nothing.